### ISSUE TITLE
win_wua.<anything> tracebacks on some computers

### ISSUE THREAD

**Issue by ooboyle** [saltstack/salt#51874]
On some of my 1000 minions, I get a traceback when using the win_wua module. I believe I'm seeing this on various Windows OS versions. Minions are all 2018.3.0. Any help would be appreciated as this is preventing me from installing packages and doing updates.

One way to reproduce the error is:

salt-call -l debug win_wua.list install=True

    [DEBUG   ] Reading configuration from c:\salt\conf\minion
    [DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
    [DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
    [DEBUG   ] Using cached minion ID from c:\salt\conf\minion_id: minion
    [DEBUG   ] Configuration file path: c:\salt\conf\minion
    [WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
    [DEBUG   ] Grains refresh requested. Refreshing grains.
    [DEBUG   ] Reading configuration from c:\salt\conf\minion
    [DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
    [DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
    [DEBUG   ] Failed to import grains show_current_loc_py3:
    Traceback (most recent call last):
      File "C:\salt\bin\lib\site-packages\salt\loader.py", line 1459, in _load_module
        mod = imp.load_module(mod_namespace, fn_, fpath, desc)
      File "c:\salt\var\cache\salt\minion\extmods\grains\show_current_loc_py3.py", line 6, in <module>
        from urllib.parse import urlparse, urlencode
    ImportError: No module named parse
    [DEBUG   ] Loading static grains from c:\salt\conf\grains
    [DEBUG   ] Connecting to master. Attempt 1 (infinite attempts)
    [DEBUG   ] Master URI: tcp://192.168.214.218:4506
    [DEBUG   ] Popen(['git', 'version'], cwd=C:\Windows\system32\config\systemprofile, universal_newlines=False, shell=None)
    [DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'minion', u'tcp://192.168.214.218:4506')
    [DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (4423)
    [DEBUG   ] Setting zmq_reconnect_ivl to '4423ms'
    [DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
    [DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'minion', u'tcp://192.168.214.218:4506', 'clear')
    [DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.214.218:4506
    [DEBUG   ] Trying to connect to: tcp://192.168.214.218:4506
    [DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
    [DEBUG   ] Decrypting the current master AES key
    [DEBUG   ] salt.crypt.get_rsa_key: Loading private key
    [DEBUG   ] salt.crypt._get_key_with_evict: Loading private key
    [DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
    [DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
    [DEBUG   ] Connecting the Minion to the Master publish port, using the URI: tcp://192.168.214.218:4505
    [DEBUG   ] salt.crypt.get_rsa_key: Loading private key
    [DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
    [DEBUG   ] Determining pillar cache
    [DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'minion', u'tcp://192.168.214.218:4506', u'aes')
    [DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'minion', u'tcp://192.168.214.218:4506')
    [DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.214.218:4506
    [DEBUG   ] Trying to connect to: tcp://192.168.214.218:4506
    [DEBUG   ] salt.crypt.get_rsa_key: Loading private key
    [DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
    [DEBUG   ] LazyLoaded jinja.render
    [DEBUG   ] LazyLoaded yaml.render
    [DEBUG   ] LazyLoaded win_wua.list
    [ERROR   ] Search Failed: Unknown Failure: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -939523085), None)
                Type='Software' or Type='Driver'
    Traceback (most recent call last):
      File "C:\salt\bin\lib\site-packages\salt\cli\caller.py", line 212, in call
        ret['return'] = func(*args, **kwargs)
      File "C:\salt\bin\lib\site-packages\salt\modules\win_wua.py", line 668, in list
        wua = salt.utils.win_update.WindowsUpdateAgent()
      File "C:\salt\bin\lib\site-packages\salt\utils\win_update.py", line 284, in __init__
        self.refresh()
      File "C:\salt\bin\lib\site-packages\salt\utils\win_update.py", line 352, in refresh
        raise CommandExecutionError(failure_code)
    CommandExecutionError: Unknown Failure: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -939523085), None)
    Error running 'win_wua.list': Unknown Failure: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -939523085), None)

    Salt Version:
               Salt: 2018.3.0

    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 10.2.1
           dateutil: 2.6.1
          docker-py: Not Installed
              gitdb: 2.0.3
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.10
            libgit2: Not Installed
            libnacl: 1.6.1
           M2Crypto: Not Installed
               Mako: 1.0.7
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.18
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.4.1
             PyYAML: 3.12
              PyZMQ: 16.0.3
               RAET: Not Installed
              smmap: 2.0.3
            timelib: 0.2.4
            Tornado: 4.5.2
                ZMQ: 4.1.6

    System Versions:
               dist:
             locale: cp1252
            machine: AMD64
            release: 7
             system: Windows
            version: 7 6.1.7601 SP1 Multiprocessor Free

Thanks!
Oliver

**Comment by twangboy** [comment]
That's an out of memory error:
```
0xC80003F3 -939523085 hrOutOfMemory The computer is out of memory 
```
from https://social.technet.microsoft.com/wiki/contents/articles/15260.windows-update-agent-error-codes.aspx

Can you look at the machine and see if there's a process using up all the memory?

**Comment by ooboyle** [comment]
twangboy:

Thanks for the reply. I get the error immediately after a reboot as well. I don't have GUI access to the device right now but running the systeminfo command via cmd.run shows:

    Total Physical Memory:     8,066 MB
    Available Physical Memory: 6,355 MB
    Virtual Memory: Max Size:  16,130 MB
    Virtual Memory: Available: 14,353 MB
    Virtual Memory: In Use:    1,777 MB

I feel like there's enough memory available.

Oliver

**Comment by ooboyle** [comment]
By process:

    Image Name                     PID Session Name        Session#    Mem Usage
    ========================= ======== ================ =========== ============
    System Idle Process              0 Services                   0         24 K
    System                           4 Services                   0     23,860 K
    smss.exe                       368 Services                   0      1,276 K
    csrss.exe                      544 Services                   0      5,316 K
    csrss.exe                      608 Console                    1      9,156 K
    wininit.exe                    616 Services                   0      5,672 K
    winlogon.exe                   668 Console                    1      8,052 K
    services.exe                   716 Services                   0     16,388 K
    lsass.exe                      732 Services                   0     14,392 K
    lsm.exe                        744 Services                   0      6,888 K
    svchost.exe                    852 Services                   0     10,292 K
    svchost.exe                    936 Services                   0      8,528 K
    LogonUI.exe                   1020 Console                    1     23,500 K
    svchost.exe                    556 Services                   0     20,944 K
    svchost.exe                    612 Services                   0    119,532 K
    svchost.exe                    760 Services                   0     13,516 K
    svchost.exe                    976 Services                   0     34,376 K
    svchost.exe                   1056 Services                   0     12,756 K
    igfxCUIService.exe            1136 Services                   0      6,792 K
    RtkAudioService64.exe         1232 Services                   0      6,404 K
    RAVBg64.exe                   1260 Console                    1     13,092 K
    svchost.exe                   1472 Services                   0     28,364 K
    spoolsv.exe                   1596 Services                   0     25,164 K
    svchost.exe                   1628 Services                   0     13,480 K
    armsvc.exe                    1892 Services                   0      4,948 K
    AERTSr64.exe                  1928 Services                   0      2,960 K
    mDNSResponder.exe             1952 Services                   0      6,964 K
    RepMgr64.exe                  1984 Services                   0     90,156 K
    RepWSC64.exe                  2012 Services                   0      6,064 K
    officeclicktorun.exe           440 Services                   0     22,200 K
    svchost.exe                   1552 Services                   0     13,204 K
    HeciServer.exe                2180 Services                   0      6,648 K
    IPROSetMonitor.exe            2216 Services                   0      5,488 K
    ir_agent.exe                  2368 Services                   0     13,428 K
    RepUtils32.exe                2496 Services                   0      3,140 K
    conhost.exe                   2540 Services                   0      4,488 K
    Jhi_service.exe               2584 Services                   0      6,052 K
    ir_agent.exe                  2592 Services                   0     55,292 K
    svchost.exe                   2660 Services                   0      4,868 K
    svchost.exe                   2708 Services                   0      4,796 K
    svchost.exe                   2736 Services                   0      4,708 K
    ssm.exe                       2800 Services                   0      5,620 K
    python.exe                    1556 Services                   0     81,768 K
    conhost.exe                    192 Services                   0      4,580 K
    svchost.exe                   2720 Services                   0      6,540 K
    svchost.exe                   3856 Services                   0     28,960 K
    python.exe                    2672 Services                   0     41,160 K
    ir_agent.exe                  3960 Services                   0     58,712 K
    ir_agent.exe                  2972 Services                   0     54,504 K
    ir_agent.exe                   212 Services                   0     48,300 K
    honeyhashx86.exe              4008 Services                   0      3,716 K
    conhost.exe                   3992 Services                   0      4,452 K
    ir_agent.exe                  2668 Services                   0     53,956 K
    IAStorDataMgrSvc.exe          2820 Services                   0     18,912 K
    LMS.exe                       2792 Services                   0      6,388 K
    GoogleCrashHandler.exe        3144 Services                   0        528 K
    GoogleCrashHandler64.exe      2932 Services                   0        588 K
    SearchIndexer.exe             1176 Services                   0     15,192 K
    UNS.exe                       3996 Services                   0     14,484 K
    WmiPrvSE.exe                  1716 Services                   0      7,312 K
    WmiPrvSE.exe                  2404 Services                   0     11,444 K
    python.exe                    2532 Services                   0     69,660 K
    cmd.exe                       2880 Services                   0      4,256 K
    tasklist.exe                  1900 Services                   0      6,884 K

**Comment by ooboyle** [comment]
I can track the apps and processes being called via my antivirus app and I see that salt-call is trying to modify memory (last line pasted below).

Event ID: 4c8e808a3ac011e9a4e90f67e93757e8 Device location: On-Premise Category: Monitored Process started: A few seconds ago Device IP address:  Device OS: Windows 7 x64 SP: 1 User Name: SYSTEM Sensor installed By: CSHS\kitchenstaff Parent name: cmd.exe Parent process ID: 3492 Parent reputation: TRUSTED_WHITE_LIST Parent reputation (applied, white database): TRUSTED_WHITE_LIST Parent SHA: db06c3534964e3fc79d2763144ba53742d7fa250ca336f4a0fe724b75aaff386 Parent command line: C:\Windows\system32\cmd.exe /c "chcp 437 > nul & salt-call -l debug win_wua.list install=True" Process name: salt-call Process ID: 3576 App reputation: COMPANY_WHITE_LIST App reputation (applied, hash reputation list): COMPANY_WHITE_LIST App MD5: d37dad3b3887be9ba1d1d898de1ea0c1 App SHA: 027a241f2ae5defc102be4b0934d914e240f0aa37ee9f8497617168191f81c9f Command line: "C:\salt\bin\python.exe" -E -s "C:\salt\bin\Scripts\salt-call" -l debug win_wua.list install=True Target Name: python.exe Target command line: python.exe TTPs: MODIFY_MEMORY_PROTECTION

**Comment by twangboy** [comment]
What's the exact command you're running?
It looks like you're running something like this:
```
salt * cmd.run 'salt-call -l debug win_wua.list install=True'
```
If that's the case, let's work with win_wua directly:
```
salt * win_wua.list install=True
```

**Comment by ooboyle** [comment]
was running:

salt <minion_name> cmd.run 'salt-call -l debug win_wua.list install=True'

and 

salt win_wua.list install=True

same result in both cases. First option shows more errors so I sometimes use it to get a different view of the issue.

**Comment by ooboyle** [comment]
Turns out it was a windows update agent issue. It needed to be reset, likely due to some sort of corruption. I did this:

net stop wuauserv
net stop cryptSvc
net stop bits
net stop msiserver
ren C:\Windows\SoftwareDistribution SoftwareDistribution.old
ren C:\Windows\System32\catroot2 Catroot2.old
net start wuauserv
net start cryptSvc
net start bits
net start msiserver

reboot

Thanks for the help. I'm glad it didn't end up being salt!

Oliver
